jsre/pretty.go

```diff
@@ -0,0 +1,220 @@
+		specialColor.Print("undefined")
+	case v.IsString():
+		s, _ := v.ToString()
+		stringColor.Printf("%q", s)
+	case v.IsBoolean():
+		b, _ := v.ToBoolean()
+		specialColor.Printf("%t", b)
+	case v.IsNaN():
+		numberColor.Printf("NaN")
+	case v.IsNumber():
+		s, _ := v.ToString()
+		numberColor.Printf("%s", s)
+	default:
+		fmt.Printf("<unprintable>")
+	}
+}
+
+func (ctx ppctx) printObject(obj *otto.Object, level int) {
+	switch obj.Class() {
+	case "Array":
+		lv, _ := obj.Get("length")
+		len, _ := lv.ToInteger()
+		if len == 0 {
+			fmt.Printf("[]")
+			return
+		}
+		if level > maxPrettyPrintLevel {
+			fmt.Print("[...]")
+			return
+		}
+		fmt.Print("[")
+		for i := int64(0); i < len; i++ {
+			el, err := obj.Get(strconv.FormatInt(i, 10))
+			if err == nil {
+				ctx.printValue(el, level+1)
+			}
+			if i < len-1 {
+				fmt.Printf(", ")
+			}
+		}
+		fmt.Print("]")
+
+	case "Object":
+		// Print values from bignumber.js as regular numbers.
+		if ctx.isBigNumber(obj) {
+			numberColor.Print(toString(obj))
+			return
+		}
+		// Otherwise, print all fields indented, but stop if we're too deep.
+		keys := ctx.fields(obj)
+		if len(keys) == 0 {
+			fmt.Print("{}")
+			return
+		}
+		if level > maxPrettyPrintLevel {
+			fmt.Print("{...}")
+			return
+		}
+		fmt.Println("{")
+		for i, k := range keys {
+			v, _ := obj.Get(k)
+			fmt.Printf("%s%s: ", ctx.indent(level+1), k)
+			ctx.printValue(v, level+1)
+			if i < len(keys)-1 {
+				fmt.Printf(",")
+			}
+			fmt.Println()
+		}
+		fmt.Printf("%s}", ctx.indent(level))
+
+	case "Function":
+		// Use toString() to display the argument list if possible.
+		if robj, err := obj.Call("toString"); err != nil {
+			functionColor.Print("function()")
+		} else {
+			desc := strings.Trim(strings.Split(robj.String(), "{")[0], " \t\n")
+			desc = strings.Replace(desc, " (", "(", 1)
+			functionColor.Print(desc)
+		}
+
+	case "RegExp":
+		stringColor.Print(toString(obj))
+
+	default:
+		if v, _ := obj.Get("toString"); v.IsFunction() && level <= maxPrettyPrintLevel {
+			s, _ := obj.Call("toString")
+			fmt.Printf("<%s %s>", obj.Class(), s.String())
+		} else {
+			fmt.Printf("<%s>", obj.Class())
+		}
+	}
+}
+
+func (ctx ppctx) fields(obj *otto.Object) []string {
+	var (
+		vals, methods []string
+		seen          = make(map[string]bool)
+	)
+	add := func(k string) {
+		if seen[k] || boringKeys[k] {
+			return
+		}
+		seen[k] = true
+		if v, _ := obj.Get(k); v.IsFunction() {
+			methods = append(methods, k)
+		} else {
+			vals = append(vals, k)
+		}
+	}
+	// add own properties
+	ctx.doOwnProperties(obj.Value(), add)
+	// add properties of the constructor
+	if cp := constructorPrototype(obj); cp != nil {
+		ctx.doOwnProperties(cp.Value(), add)
+	}
+	sort.Strings(vals)
+	sort.Strings(methods)
+	return append(vals, methods...)
+}
+
+func (ctx ppctx) doOwnProperties(v otto.Value, f func(string)) {
+	Object, _ := ctx.vm.Object("Object")
+	rv, _ := Object.Call("getOwnPropertyNames", v)
+	gv, _ := rv.Export()
```

Comment:
The easiest way to break this is to evaluate `Object.getOwnPropertyNames = null` at the console.
Printing any object after that crashes the console. I don't think this is worth fixing though. The best
thing we could do in such a situation would be to print something like "can't print because you broke Object".